GitVersionExe/GitPreparer.cs

```diff
@@ -20,31 +21,74 @@ public bool IsDynamicGitRepository
 
         public string DynamicGitRepositoryPath { get; private set; }
 
-        public string Prepare()
+        public void InitialiseDynamicRepositoryIfNeeded()
         {
-            var gitPath = arguments.TargetPath;
+            if (string.IsNullOrWhiteSpace(arguments.TargetUrl)) return;
 
-            if (!string.IsNullOrWhiteSpace(arguments.TargetUrl))
+            var targetPath = CalculateTemporaryRepositoryPath(arguments.TargetUrl);
+            DynamicGitRepositoryPath = CreateDynamicRepository(targetPath, arguments.Authentication, arguments.TargetUrl, arguments.TargetBranch);
+        }
+
+        string CalculateTemporaryRepositoryPath(string targetUrl)
+        {
+            var userTemp = Path.GetTempPath();
+            var repositoryName = targetUrl.Split('/', '\\').Last().Replace(".git", string.Empty);
+            var possiblePath = Path.Combine(userTemp, repositoryName);
+
+            // Verify that the existing directory is ok for us to use
+            if (Directory.Exists(possiblePath))
             {
-                gitPath = GetGitInfoFromUrl();
+                if (!GitRepoHasMatchingRemote(possiblePath, targetUrl))
+                {
+                    var i = 1;
+                    var originalPath = possiblePath;
+                    bool possiblePathExists;
+                    do
+                    {
+                        possiblePath = string.Concat(originalPath, "_", i++.ToString());
+                        possiblePathExists = Directory.Exists(possiblePath);
+                    } while (possiblePathExists && !GitRepoHasMatchingRemote(possiblePath, targetUrl));
+                }
             }
 
-            return GitDirFinder.TreeWalkForGitDir(gitPath);
+            return possiblePath;
         }
 
-        string GetGitInfoFromUrl()
+        static bool GitRepoHasMatchingRemote(string possiblePath, string targetUrl)
         {
-            var gitRootDirectory = Path.Combine(arguments.TargetPath, "_dynamicrepository");
-            var gitDirectory = Path.Combine(gitRootDirectory, ".git");
-            if (Directory.Exists(gitRootDirectory))
+            try
+            {
+                using (var repository = new Repository(possiblePath))
```

Comment:
Not sure (haven't looked into the code), but shouldn't we do a fetch first to check if there are new remotes (for example, for PR)?